### PR TITLE
convert some CTU hydro routines to C++

### PR DESCRIPTION
**Pull request by zingale**
## PR summary

This converts the CTU consup for hydro (not radiation) , the flattening routines, and src_to_prim to C++

## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite
- [ ] all functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
all CPU tests pass
need to do the GPU tests too

**Comment by zingale** [comment]
GPU works for everything except wdmerger.  This is a hybrid hydro issue.  We get:
```
0: copyout Memcpy (host=0x7ffecdac3520, dev=0x7f2af7025400, size=24) FAILED: 719(unspecified launch failure)```

**Comment by zingale** [comment]
The issue is the call to `position_c`

**Comment by zingale** [comment]
GPU tests now pass:
http://groot.astro.sunysb.edu/Castro/test-suite/gpu/2020-02-10-001/index.html

CPU tests pass too (note: the failures are because of changes on development not merged into this):
http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-02-10-005/index.html

this is ready to be merged

File: Source/hydro/flatten.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    // x-direction flattening coef
+
+    Real dp = q(i+1,j,k,pres_comp) - q(i-1,j,k,pres_comp);
+
+    int ishft = dp > 0.0 ? 1 : -1;
+
+    Real denom = std::max(small_pres, std::abs(q(i+2,j,k,pres_comp) - q(i-2,j,k,pres_comp)));
+    Real zeta = std::abs(dp) / denom;
+    Real z = std::min(1.0, std::max(0.0, dzcut * (zeta - zcut1)));
```
Comment (maxpkatz) [inline Source/hydro/flatten.cpp]: std::min and std::max are unavailable in CUDA device code. Please use amrex::min and amrex::max (they're defined in AMReX_Algorithm.H, and should be readily available here).

File: Source/hydro/Castro_ctu_hydro.cpp [hunk 2]
```diff
@@ -176,22 +176,37 @@ Castro::construct_ctu_hydro_source(Real time, Real dt)
...
-        ca_uflatten(AMREX_INT_ANYD(obx.loVect()), AMREX_INT_ANYD(obx.hiVect()),
-                    BL_TO_FORTRAN_ANYD(q[mfi]),
-                    BL_TO_FORTRAN_ANYD(flatn), QPRES+1);
+        uflatten(obx, q_arr, flatg_arr, QPTOT);
+
+        AMREX_PARALLEL_FOR_3D(obx, i, j, k,
+        {
+          flatn_arr(i,j,k) = flatn_arr(i,j,k) * flatg_arr(i,j,k);
+
+          if (Radiation::flatten_pp_threshold > 0.0) {
```
Comment (maxpkatz) [inline Source/hydro/Castro_ctu_hydro.cpp]: We cannot directly read class member variables in the lambdas on the device. For now we'll need to make a local copy of this variable outside the parallel for.

File: Source/hydro/Castro_ctu_hydro.cpp [hunk 3]
```diff
@@ -176,22 +176,39 @@ Castro::construct_ctu_hydro_source(Real time, Real dt)
...
+
+        AMREX_PARALLEL_FOR_3D(obx, i, j, k,
+        {
+          flatn_arr(i,j,k) = flatn_arr(i,j,k) * flatg_arr(i,j,k);
+
+          if (flatten_pp_thresh > 0.0) {
+            if ( q_arr(i-1,j,k,QU) + q_arr(i,j-1,k,QV) + q_arr(i,j,k-1,QW) >
+                 q_arr(i+1,j,k,QU) + q_arr(i,j+1,k,QV) + q_arr(i,j,k+1,QW) ) {
+
+              if (q_arr(i,j,k,QPRES) < Radiation::flatten_pp_threshold * q_arr(i,j,k,QPTOT)) {
```
Comment (maxpkatz) [inline Source/hydro/Castro_ctu_hydro.cpp]: Same comment for this one :-)

**Pull request merged**